src/fmpz_vec/dot.c

```diff
@@ -11,26 +9,394 @@
+    {
+        fmpz ca;
+        mp_limb_t atmp;
+        mp_srcptr ap;
+        mp_size_t an;
+        int aneg;
+
+        ca = *initial;
+        FMPZ_GET_MPN(ap, an, aneg, atmp, ca);
+
+        if (an <= 2)
+        {
+            s0 = ap[0];
+            if (an == 2)
+                s1 = ap[1];
+
+            if (aneg ^ subtract)
+                sub_dddmmmsss(s2, s1, s0, 0, 0, 0, 0, s1, s0);
+        }
+        else
+        {
+            if (an > INITIAL_ALLOC)
+            {
+                new_alloc = an + 4;
+
+                tmp_heap = flint_malloc(3 * (new_alloc + 2) * sizeof(mp_limb_t));
+
+                t = tmp_heap;
+                pos = t + (new_alloc + 2);
+                neg = pos + (new_alloc + 2);
+
+                alloc = new_alloc;
+            }
+
+            if (aneg ^ subtract)
+            {
+                flint_mpn_copyi(neg, ap, an);
+                negn = an;
+            }
+            else
+            {
+                flint_mpn_copyi(pos, ap, an);
+                posn = an;
+            }
+        }
+    }
+
     for (i = 0; i < len; i++)
-        fmpz_addmul(c, vec1 + i, vec2[i] + offset);
+    {
+        fmpz ca, cb;
+        mp_limb_t atmp, btmp;
+        mp_srcptr ap, bp;
+        mp_size_t an, bn;
+        mp_limb_t cy;
+        int aneg, bneg;
+
+        ca = a[i];
+        if (ca == 0)
+            continue;
+
+        cb = reverse ? b[len - 1 - i] : b[i];
+        if (cb == 0)
+            continue;
+
+        if (!COEFF_IS_MPZ(ca) && !COEFF_IS_MPZ(cb))
+        {
+            mp_limb_t hi, lo;
+            smul_ppmm(hi, lo, ca, cb);
+            add_sssaaaaaa(s2, s1, s0, s2, s1, s0, FLINT_SIGN_EXT(hi), hi, lo);
+            continue;
+        }
+
+        FMPZ_GET_MPN(ap, an, aneg, atmp, ca);
+        FMPZ_GET_MPN(bp, bn, bneg, btmp, cb);
+        tn = an + bn;
+
+        if (tn > alloc)
+        {
+            mp_ptr p1, p2, p3;
+
+            new_alloc = FLINT_MAX(3 * alloc / 2, tn + 4);
+
+            p1 = flint_malloc(3 * (new_alloc + 2) * sizeof(mp_limb_t));
+            p2 = p1 + (new_alloc + 2);
+            p3 = p2 + (new_alloc + 2);
+
+            flint_mpn_copyi(p2, pos, posn);
+            flint_mpn_copyi(p3, neg, negn);
+            t = p1;
+            pos = p2;
+            neg = p3;
+
+            FLINT_SWAP(mp_ptr, tmp_heap, p1);
+
+            if (p1 != NULL)
+                flint_free(p1);
+
+            alloc = new_alloc;
+        }
+
+        if (an < bn)
+        {
+            FLINT_SWAP(mp_srcptr, ap, bp);
+            FLINT_SWAP(mp_size_t, an, bn);
+        }
+
+        if (bn == 1)
+        {
+            mp_limb_t b0 = bp[0];
+
+            if (aneg ^ bneg)
+                MPN_ADDMUL_1(neg, negn, ap, an, b0);
+            else
+                MPN_ADDMUL_1(pos, posn, ap, an, b0);
+            continue;
+        }
+
+        if (ap == bp && an == bn)
+        {
+            flint_mpn_sqr(t, ap, an);
+            cy = t[tn - 1];
+        }
```

Comment:
I think it is better to skip this for now. In case we want to calculate `x \dot x`, we should implement a separate function as that should be faster due to less branching.

Reply:
Yes, there could be a separate function for this, but there are also situations where only some entries in a dot product happen to be squares. For example, when squaring matrices.

Reply:
True.